inttest/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,260 @@
+// - Check the status persists between upgrades.
+// - Check the k0smotron cluster is still accessible between upgrades.
+// - Check the status of the k0smotron cluster is Ready and with it its related resources after upgrade.
+func (s *UpgradeSuite) TestK0smotronUpgrade() {
+	s.T().Log("starting k0s")
+	s.Require().NoError(s.InitController(0, "--disable-components=konnectivity-server,metrics-server"))
+	s.Require().NoError(s.RunWorkers())
+
+	kc, err := s.KubeClient(s.ControllerNode(0))
+	s.Require().NoError(err)
+	rc, err := s.GetKubeConfig(s.ControllerNode(0))
+	s.Require().NoError(err)
+
+	err = s.WaitForNodeReady(s.WorkerNode(0), kc)
+	s.NoError(err)
+
+	s.T().Log("deploying stable k0smotron operator")
+	s.Require().NoError(util.InstallStableK0smotronOperator(s.Context(), kc, rc))
+	s.Require().NoError(common.WaitForDeployment(s.Context(), kc, "k0smotron-controller-manager", "k0smotron"))
+
+	s.T().Log("deploying k0smotron cluster")
+	s.createK0smotronCluster(s.Context(), kc)
+	s.Require().NoError(common.WaitForStatefulSet(s.Context(), kc, "kmc-kmc-test", "kmc-test"))
+
+	pod, err := kc.CoreV1().Pods("kmc-test").Get(s.Context(), "kmc-kmc-test-0", metav1.GetOptions{})
+	s.Require().NoError(err)
+
+	// We create state to subsequently validate that it persists between upgrades
+	s.addStateInMountedPath(s.Context(), pod.Spec.Containers[0].VolumeMounts[1].MountPath, kc, rc)
+
+	s.T().Log("deploying development k0smotron operator")
+	s.Require().NoError(s.ImportK0smotronImages(s.Context()))
+	s.Require().NoError(util.ApplyFromYAML(s.Context(), kc, rc, os.Getenv("K0SMOTRON_INSTALL_YAML")))
+	s.Require().NoError(util.WaitForRolloutCompleted(s.Context(), kc, "k0smotron-controller-manager", "k0smotron"))
+
+	s.forceControllerRecreation(s.Context(), pod.Name, kc)
+	s.checkStatePersists(s.Context(), pod.Spec.Containers[0].VolumeMounts[1].MountPath, kc, rc)
+
+	s.T().Log("Starting portforward")
+	fw, err := util.GetPortForwarder(rc, "kmc-kmc-test-0", "kmc-test", 30443)
+	s.Require().NoError(err)
+
+	go fw.Start(s.Require().NoError)
+	defer fw.Close()
+
+	<-fw.ReadyChan
+
+	localPort, err := fw.LocalPort()
+	s.Require().NoError(err)
+	kmcKC, err := util.GetKMCClientSet(s.Context(), kc, "kmc-test", "kmc-test", localPort)
+	s.Require().NoError(err)
+
+	_, err = kmcKC.CoreV1().Namespaces().Get(s.Context(), "test-ns-cm", metav1.GetOptions{})
+	s.Require().NoError(err)
+
+	result, err := kc.RESTClient().Get().AbsPath("/apis/k0smotron.io/v1beta1/namespaces/kmc-test/clusters/kmc-test").DoRaw(s.Context())
+	s.Require().NoError(err)
+
+	// If the status of the Owner is correct, the status of all resources
+	// that have OwnerRef to this resource should also be correct.
+	var kmc km.Cluster
+	err = yaml.Unmarshal(result, &kmc)
+	s.Require().NoError(err)
+	s.Require().True(kmc.Status.Ready)
+}
+
+func (s *UpgradeSuite) checkStatePersists(ctx context.Context, mountedPath string, kc *kubernetes.Clientset, rc *rest.Config) {
```

Comment:
We also need to check the cluster state in etcd, eg create some object using API before the upgrade and check if it still exists.

Reply:
done 👍